pkg/cmd/pulumi/plugin/plugin_prune.go

```diff
@@ -0,0 +1,293 @@
+				key = fmt.Sprintf("%s|%s", plugin.Kind, plugin.Name)
+			} else {
+				// Group by kind, name, and major version
+				key = fmt.Sprintf("%s|%s|v%d", plugin.Kind, plugin.Name, plugin.Version.Major)
+			}
+		} else {
+			// If no version, just use kind and name
+			key = fmt.Sprintf("%s|%s", plugin.Kind, plugin.Name)
+		}
+
+		groups[key] = append(groups[key], plugin)
+	}
+
+	// For each group, identify plugins to remove (all but the latest version)
+	toRemove := make([]workspace.PluginInfo, 0, len(plugins))
+	toKeep := make([]workspace.PluginInfo, 0, len(plugins))
+	var totalSizeRemoved uint64
+
+	for _, group := range groups {
+		if len(group) <= 1 {
+			// Only one version, keep it
+			toKeep = append(toKeep, group[0])
+			continue
+		}
+
+		// Sort plugins by version in descending order (newer versions first).
+		// Plugins with nil versions are treated as lower priority and are sorted to
+		// the end of the list. This ensures plugins with defined versions are preferred
+		// when determining which version to keep.
+		sort.Slice(group, func(i, j int) bool {
+			// If either version is nil, keep the one with a version
+			if group[i].Version == nil {
+				return false
+			}
+			if group[j].Version == nil {
+				return true
+			}
+			// Otherwise sort by version (newer versions first)
+			return group[i].Version.GT(*group[j].Version)
+		})
+
+		// Keep only the latest version in each group
+		toKeep = append(toKeep, group[0])
+
+		// Remove the rest
+		for i := 1; i < len(group); i++ {
+			toRemove = append(toRemove, group[i])
+			totalSizeRemoved += group[i].Size
+		}
+	}
+
+	if len(toRemove) == 0 {
+		cmd.diag.Infof(
+			diag.Message("", "no plugins found to prune"))
+		return nil
+	}
+
+	// Confirm that the user wants to do this (unless --yes was passed)
+	if !cmd.dryRun && !cmd.yes {
+		var suffix string
+		if len(toRemove) != 1 {
+			suffix = "s"
+		}
+		fmt.Fprint(
+			cmd.Stdout,
+			opts.Color.Colorize(
+				fmt.Sprintf("%sThis will remove %d plugin%s from the cache, reclaiming %s:%s\n",
+					colors.SpecAttention, len(toRemove), suffix,
+					humanize.Bytes(totalSizeRemoved), colors.Reset)))
+
+		fmt.Fprintln(cmd.Stdout, "Plugins to remove:")
+		// Sort by name and version for a nice display
+		sortPluginsByNameKindVersion(toRemove)
+
+		for _, plugin := range toRemove {
+			versionStr := "n/a"
+			if plugin.Version != nil {
+				versionStr = plugin.Version.String()
+			}
+			fmt.Fprintf(cmd.Stdout, "    %s %s v%s (%s)\n",
+				plugin.Kind, plugin.Name, versionStr, humanize.Bytes(plugin.Size))
+		}
+
+		fmt.Fprintln(cmd.Stdout, "\nPlugins to keep:")
+		// Sort the kept plugins as well
+		sortPluginsByNameKindVersion(toKeep)
+
+		displayedKinds := make(map[string]bool)
+		for _, plugin := range toKeep {
+			key := fmt.Sprintf("%s|%s", plugin.Kind, plugin.Name)
+			if displayedKinds[key] {
+				continue // Only show one version of each kind|name combo to avoid long lists
+			}
```

Comment:
This is quite strange.  I'd expect us to have a fixed limit here or something, it feels kinda random to look at the list of plugins kept and not see some versions there that I would have expected to be there.